Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -855,6 +909,313 @@ lemma addY_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation
   erw [addY, negY_of_Z_ne_zero <| addZ_ne_zero_of_X_ne hx, addX_of_Z_ne_zero hP hQ hPz hQz hx,
     negAddY_of_Z_ne_zero hP hQ hPz hQz hx, Affine.addY]
 
+variable (W') in
+/-- The coordinates of the addition of two distinct point representatives. -/
+noncomputable def addXYZ (P Q : Fin 3 → R) : Fin 3 → R :=
+  ![W'.addX P Q, W'.addY P Q, addZ P Q]
+
+lemma addXYZ_smul (P Q : Fin 3 → R) (u v : Rˣ) :
+    W'.addXYZ (u • P) (v • Q) = (u * v) ^ 2 • W'.addXYZ P Q := by
+  rw [addXYZ, addX_smul, addY_smul, addZ_smul]
+  rfl
+
+lemma addXYZ_of_Z_eq_zero_left' {P Q : Fin 3 → R} (hP : W'.Equation P) (hPz : P z = 0) :
+    W'.addXYZ P Q = ![(P x * Q z) ^ 2 * Q x, (P x * Q z) ^ 3 * Q y, P x * Q z * Q z] := by
+  rw [addXYZ, addX_of_Z_eq_zero_left hPz, addY_of_Z_eq_zero_left hP hPz, addZ_of_Z_eq_zero_left hPz]
+
+lemma addXYZ_of_Z_eq_zero_left {P Q : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z = 0)
+    (hQz : Q z ≠ 0) :
+    W.addXYZ P Q = (Units.mk0 _ (X_ne_zero_of_Z_eq_zero hP hPz) * Units.mk0 _ hQz) • Q :=
+  addXYZ_of_Z_eq_zero_left' hP.left hPz
+
+lemma addXYZ_of_Z_eq_zero_right' {P Q : Fin 3 → R} (hQ : W'.Equation Q) (hQz : Q z = 0) :
+    W'.addXYZ P Q = ![(-(Q x * P z)) ^ 2 * P x, (-(Q x * P z)) ^ 3 * P y, -(Q x * P z) * P z] := by
+  rw [addXYZ, addX_of_Z_eq_zero_right hQz, addY_of_Z_eq_zero_right hQ hQz,
+    addZ_of_Z_eq_zero_right hQz]
+
+lemma addXYZ_of_Z_eq_zero_right {P Q : Fin 3 → F} (hQ : W.Nonsingular Q) (hPz : P z ≠ 0)
+    (hQz : Q z = 0) :
+    W.addXYZ P Q = -(Units.mk0 _ (X_ne_zero_of_Z_eq_zero hQ hQz) * Units.mk0 _ hPz) • P :=
+  addXYZ_of_Z_eq_zero_right' hQ.left hQz
+
+lemma addXYZ_of_Y_ne {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 ≠ Q y * P z ^ 3) :
+    W.addXYZ P Q =
+      -(Units.mk0 _ (sub_ne_zero_of_ne hy) / (Units.mk0 _ hPz * Units.mk0 _ hQz)) • ![1, 1, 0] := by
```

Comment:
As I said I think it's reasonable to generalize the scalar multiplication to get rid of Units.mk0. This division in `Units F` could be replaced with division in the field F.

Restricting to scalar multiplication by units forces you to state less general results, which I think is bad.

Reply:
Oh I'm totally for this, I just wasn't sure before whether I should combine the changes with this PR, but I guess I will.

Reply:
I have extended the scalar multiplication to include non-units in the latest commit, but I think that the equivalence relation must still be on the units e.g. to prove `nonsingular_smul`, so much of the code will now require `IsUnit u`. Overall this increases the code length, because now we have to add the `IsUnit` lemmas, but I think once we refactor `Nonsingular` into the correct definition over a larger class of rings e.g. valuation rings, this will save us some work.

Reply:
Thanks! Yes the equivalence relation needs a **group** action, and only units have inverses.

It seems my suggestion to replace the `Units.mk0` is what makes the code longer, which I regret. By using `Units.mk0` you essentially bundled the nonzero proof inside the statement of the lemma, so you no longer need the nonzero proofs when applying the lemma. Since these lemmas seem to be auxiliary and applied at multiple occasions, it may be wise to keep the original `Units.mk0` statement, and only generalize selected lemmas like in [my branch](https://github.com/leanprover-community/mathlib4/compare/EllipticCurve.Jacobian...EllipticCurve.Jacobian+) which don't increase code length.

I'm not sure it's very useful to generalize from `ne_zero` to `isUnit` for a larger class of rings, and it may be a tedious task to figure out which results generalize. (It might be more useful in the projective case, we'll see.)

Reply:
We'll stick with this for now, even if it's longer.